lib/integrations/awsoidc/deployservice_update.go

```diff
@@ -0,0 +1,244 @@
+			"Add the following tags to allow Teleport to manage this service: %s", aws.ToString(service.ServiceName), ownershipTags)
+	}
+	// If the LaunchType is the required one, than we can update the current Service.
+	// Otherwise we have to delete it.
+	if service.LaunchType != ecsTypes.LaunchTypeFargate {
+		return nil, trace.Errorf("ECS Service %q already exists but has an invalid LaunchType %q. Delete the Service and try again.", aws.ToString(service.ServiceName), service.LaunchType)
+	}
+
+	return &service, nil
+}
+
+func getManagedTaskDefinition(ctx context.Context, clt DeployServiceClient, taskDefinitionName string, ownershipTags AWSTags) (*ecsTypes.TaskDefinition, error) {
+	describeTaskDefinitionOut, err := clt.DescribeTaskDefinition(ctx, &ecs.DescribeTaskDefinitionInput{
+		TaskDefinition: aws.String(taskDefinitionName),
+		Include:        []ecsTypes.TaskDefinitionField{ecsTypes.TaskDefinitionFieldTags},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if !ownershipTags.MatchesECSTags(describeTaskDefinitionOut.Tags) {
+		return nil, trace.Errorf("ECS Task Definition %q already exists but is not managed by Teleport. "+
+			"Add the following tags to allow Teleport to manage this task definition: %s", taskDefinitionName, ownershipTags)
+	}
+	return describeTaskDefinitionOut.TaskDefinition, nil
+}
+
+func getTaskDefinitionTeleportImage(taskDefinition *ecsTypes.TaskDefinition) (string, error) {
+	if len(taskDefinition.ContainerDefinitions) != 1 {
+		return "", trace.BadParameter("expected 1 task container definition, but got %d", len(taskDefinition.ContainerDefinitions))
+	}
+	return aws.ToString(taskDefinition.ContainerDefinitions[0].Image), nil
+}
+
+// updateServiceOrRollback attempts to update the service with the specified task definition.
+// The service will be rolled back if the service fails to become healthy.
+func updateServiceOrRollback(ctx context.Context, clt DeployServiceClient, service *ecsTypes.Service, taskDefinition *ecsTypes.TaskDefinition) (*ecsTypes.Service, error) {
+	// Update service with new task definition
+	updateServiceOut, err := clt.UpdateService(ctx, generateServiceWithTaskDefinition(service, aws.ToString(taskDefinition.TaskDefinitionArn)))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	serviceStableWaiter := ecs.NewServicesStableWaiter(clt)
+	err = serviceStableWaiter.Wait(ctx, &ecs.DescribeServicesInput{
+		Services: []string{aws.ToString(updateServiceOut.Service.ServiceName)},
+		Cluster:  updateServiceOut.Service.ClusterArn,
+	}, waitDuration)
+	if err == nil {
+		return updateServiceOut.Service, nil
+	}
+
+	// If the service fails to reach a stable state within the allowed wait time,
+	// then rollback service with previous task definition
+	rollbackServiceOut, rollbackErr := clt.UpdateService(ctx, generateServiceWithTaskDefinition(service, aws.ToString(service.TaskDefinition)))
+	if rollbackErr != nil {
+		return nil, trace.Wrap(err, "failed to rollback service: %v", err)
```

Comment:
Should probably use `trace.NewAggregate` to include both `err` and `rollbackErr` here.